pkg/operator/controller/canary/controller.go

```diff
@@ -306,17 +311,45 @@ func (r *reconciler) startCanaryRoutePolling(stop <-chan struct{}) error {
 	return nil
 }
 
-func (r *reconciler) setCanaryFailingStatusCondition() error {
+func (r *reconciler) setCanaryFailingStatusCondition(errors []error) error {
+	errorStrings := deduplicateErrorStrings(errors)
+	if len(errorStrings) > 3 {
+		errorStrings = errorStrings[len(errorStrings)-3:]
+	}
 	cond := operatorv1.OperatorCondition{
 		Type:    ingresscontroller.IngressControllerCanaryCheckSuccessConditionType,
 		Status:  operatorv1.ConditionFalse,
 		Reason:  "CanaryChecksRepetitiveFailures",
-		Message: "Canary route checks for the default ingress controller are failing",
+		Message: fmt.Sprintf("Canary route checks for the default ingress controller are failing. Last %d error messages:\n%s", len(errorStrings), strings.Join(errorStrings, "\n")),
 	}
 
 	return r.setCanaryStatusCondition(cond)
 }
 
+func deduplicateErrorStrings(errors []error) []string {
+	encountered := map[string]int{}
+	ret := make([]string, 0)
+	// Tally all occurrences of each error message, and create a list of error
+	// messages as you go so that the order is preserved
+	for i := range errors {
+		if encountered[errors[i].Error()] >= 1 {
+			encountered[errors[i].Error()] += 1
+			continue
+		}
+		encountered[errors[i].Error()] = 1
+		ret = append(ret, errors[i].Error())
+	}
+	// Now that all error messages have been tallied, add a comment to each
+	// message that was seen more than once, so the reader can tell how often
+	// each error has occurred
+	for i := range ret {
+		if encountered[ret[i]] > 1 {
+			ret[i] = fmt.Sprintf("[Seen %d times] %s", encountered[ret[i]], ret[i])
```

Comment:
Is `[Seen %d times]` used in other places? I looked around for other examples of this, and Kubernetes uses:
`(x3 over 70s)` in their describe status.

I think `[Seen %d times]` is pretty descriptive, and trying to track timestamps is a bit much (though would be pretty neat), but just was curious if you found any other examples of this?

Reply:
I haven't seen the `[Seen %d times]` format used elsewhere; I was thinking of the `(x3 over 70s)` style when I implemented it, but I guess my memory was slightly off :)

I don't think we keep timestamps of the errors at the moment, and I'm not sure how feasible it would be to implement in this bug fix. Let me take another look, though. If we do have timestamps (or if it would be easy to add), it wouldn't be too hard to track count, the first timestamp, and the last timestamp as it iterates over the list, and that would be enough for the "times over duration" style

Reply:
Yea let me know, i don't really care all that much about the exact message, but if it's reasonable to do, that'd be a bit more consistent and kind of neat.

Reply:
I took another look and while I think it would be possible to track timestamps, I don't want to over-complicate this. For now I'd prefer to leave this as-is.